README
XML::Validate v1.024

(c) BBC 2005. This program is free software; you can redistribute it and/or
modify it under the GNU GPL.

See the file COPYING in this distribution, or http://www.gnu.org/licenses/gpl.txt

QUICK START:

Install XML::Validate by unpacking the tarball and running the following
commands in the source directory:

    perl Makefile.PL
    make
    make test
    make install
    Then delete the source directory tree since it's no longer needed.

    run 'perldoc XML::Validate' to read the full documentation.

AVAILABLE BACKENDS:

Backends for the following validators are included with this distribution:

    LibXML (& you'll need XML::LibXML)
    Xerces (& you'll need XML::Xerces)
    MSXML  (for Win32 platforms, needs Win32::OLE)

The default order of preference if all 3 are available is: Xerces, MSXML, LibXML

DEPENDENCIES:

Other than the modules for use with the backends, the XML::Validate modules themselves
don't depend on other CPAN distributions.  However the unit tests distributed with
them use Test::Assertions and Log::Trace (as indicated in the Makefile.PL).

Changes
Fri Apr  7 13:08:01 2006 - 1.024
    Minor modifications to t/validate.t and t/libxml.t to allow to work in a Windows environment.

Fri Apr  7 10:50:53 2006 - 1.023
    Initial CPAN release.

Fri Apr  1 18:45:47 2005 - 1.019
    Initial release to CPAN.




scripts/validxml.pl
#!/usr/bin/perl -w

use strict;
use XML::Validate qw();
use Getopt::Long qw();
require Log::Trace;

use vars qw($VERSION);
($VERSION) = ('$Revision: 1.9 $' =~ /([\d\.]+)/ );

my $assert_invalid = 0;
my $help = 0;
my $tracing = 0;
my $deep_tracing = 0;
my $backend = 'BestAvailable';
Getopt::Long::GetOptions(
				't' => \$tracing,
				'T' => \$deep_tracing,
				'assert-invalid' => \$assert_invalid,
				'backend:s' => \$backend,
				'help' => \$help
			);

import Log::Trace 'print' if $tracing;
import Log::Trace 'print' => { Deep => 1 } if $deep_tracing;

my @files;
while (my $mask = shift) {
	push @files, glob($mask);
}
if ($help || @files < 1) {
	require Pod::Usage;
	Pod::Usage::pod2usage(-verbose => 2);
}

my $validator = new XML::Validate(Type => $backend);

my %errors;

for my $file (@files) {
	open(FH,$file) || die "Unable to open file handle FH for file '$file': $!\n";
	local $/ = undef;
	my $xml = <FH>;
	close(FH) || warn "Unable to close file handle FH for file '$file': $!\n";

	if (my $tree = $validator->validate($xml)) {
		$errors{$file} = '';
	} else {
		if ($validator->last_error()->{message}) {
			$errors{$file} = sprintf("%s at %d:%d",@{$validator->last_error()}{'message','line','column'});
		} else{
			$errors{$file} = 'Unknown error';
		}
		if (length($errors{$file}) > 0) {
			$errors{$file} =~ s/(\n|\r|\cM)/  /gi;
		}
	}
}

print "1..".@files."\n";
for my $file (@files) {
	if ($errors{$file}) {
		print(($assert_invalid ? '' : 'not ').
				"ok - $file - $errors{$file}\n");
	} else {
		print(($assert_invalid ? 'not ' : '').
				"ok - $file\n");
	}
}


=pod

=head1 NAME

validxml - Command-line driver for XML::Validate.

=head1 SYNOPSIS

    validxml *.xml
    validxml --assert-invalid *.xml

=head1 DESCRIPTION

Command-line driver for XML::Validate using the 'BestAvailable' processing
type.

=head2 Commandline Options

=over 4

=item --assert-invalid

Swap the ok/not ok so invalid things are OK - still output the validation
error) - this is useful for schema "unit tests".

=item --backend [validator type]

Specify an C<XML::Validate> backend to use (e.g LibXML, Xerces). Defaults to
I<BestAvailable>.

=back

=head2 Output

Output is Test::Harness compatible.

    1..scalar @files
    ok - filename/not ok - filename - validation error

=head1 VERSION

$Revision: 1.9 $

$Id: validxml.pl,v 1.9 2006/03/17 11:45:11 colinr Exp $

=head1 AUTHOR

Nicola Worthington

$Author: colinr $

=head1 COPYRIGHT

(c) BBC 2005. This program is free software; you can redistribute it and/or modify it under the GNU GPL.
See the file COPYING in this distribution, or http://www.gnu.org/licenses/gpl.txt

=cut

__END__



lib/XML/Validate/MSXML.pm
# XML::Validate::MSXML

package XML::Validate::MSXML;

use strict;
use Win32::OLE;
use XML::Validate::Base;
use vars qw( $VERSION @ISA $MSXML_VERSION);

@ISA = qw( XML::Validate::Base );
$VERSION = sprintf'%d.%03d', q$Revision: 1.17 $ =~ /: (\d+)\.(\d+)/;

use constant XSI_NS => 'xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"';

my $VALID_OPTIONS = { strict_validation => 1 };

sub new {
	my $class = shift;
	my %options = @_;
	my $self = {};
	bless ($self, $class);
	
	$self->clear_errors();
	$self->set_options(\%options,$VALID_OPTIONS);
	
	DUMP("Instantiating XML::Validate::MSXML", $self);
	
	return $self;
}

sub version {
	eval {create_doc_and_cache()};
	return $MSXML_VERSION;
}

sub validate {
	my ($self, $xml) = @_;
	
	$self->clear_errors();
	$self->{dom} = undef;
	
	die "validate called with no data to validate\n" unless defined $xml and length $xml > 0;

	DUMP("the xml to validate : $xml ", $self);

	my ($msxml,$msxmlcache) = create_doc_and_cache();

	$msxml->{async} = 0;
	$msxml->{validateOnParse} = 0;
	$msxml->{resolveExternals} = 1;
	
	TRACE("Starting to parse");
	$msxml->LoadXML($xml);
	TRACE("Parsed the document");
	
	my $xmlroot = $msxml->{documentElement};
	
	if ($msxml->parseError()->{errorCode} != 0) {
		TRACE("XML Parse Error (not syntactically valid)");
		my $error = $msxml->parseError();
		DUMP("Error", $error);
		$self->add_error({
			message => $error->{reason},
			line    => $error->{line},
			column  => $error->{linepos}
		});
		return;
	}
	
	if ($self->options->{strict_validation}) {
		TRACE("XML Syntactically valid");
		load_schemas($msxml, $msxmlcache);
		TRACE("Validate against schema/DTD ");
		if ($msxml->{doctype} || $msxml->{schemas}) {
			#DUMP($msxml->{doctype}, $msxml->{schemas});
			$msxml->{validateOnParse} = 1;
			$msxml->LoadXML($xml);
			my $error = $msxml->parseError();
			if ($error->{errorCode} != 0) {
				$self->add_error({
					message => $error->{reason},
					line    => $error->{line},
					column  => $error->{linepos}
				});
				return;
			}
			$error = $msxml->validate();
			if ($error->{errorCode} != 0) {
				$self->add_error({
					message => $error->{reason},
					line    => $error->{line},
					column  => $error->{linepos}
				});
				return;
			}
		} else {
			# If there is nothing to validate against, treat it as valid.
			TRACE("No doctype or schema");
		}
	}

	#Valid
	$self->{dom} = $msxml;
	return 1;
}

sub last_dom {
	my $self = shift;
	return $self->{dom};
}

sub load_schemas {
	my ($xml, $schema_cache) = @_;
	
	my %schemas;
	
	$xml->setProperty('SelectionNamespaces', XSI_NS);
	my $no_ns_schema_xpath = q[//*/@xsi:noNamespaceSchemaLocation];
	my $no_ns_schema_nodes = $xml->{documentElement}->selectNodes($no_ns_schema_xpath);
	
	for (my $i=0; $i < $no_ns_schema_nodes->{length}; $i++){
		my $schema_txt = $no_ns_schema_nodes->item($i)->{text};
		my %add_schemas = ('', $schema_txt);
		%schemas = (%schemas, %add_schemas);
	}
	
	my $schema_xpath = q[//*/@xsi:schemaLocation];
	my $schema_location_nodes = $xml->{documentElement}->selectNodes($schema_xpath);

	for (my $i=0; $i < $schema_location_nodes->{length}; $i++){
		my $schema_txt = $schema_location_nodes->item($i)->{text};
		my %add_schemas = split ' ', $schema_txt;
		%schemas = (%schemas, %add_schemas);
	}
	
	while (my ($ns, $schema) = each %schemas) {
		TRACE("Loading schema [$ns] -> $schema");
		$schema_cache->add($ns, $schema);
		if (my $schema_error = Win32::OLE::LastError()) {
			$schema_error =~ s/OLE exception .*\n\n//m;
			$schema_error =~ s/Win32::OLE.*//s;
			die $schema_error;
		}
	}
	$xml->{schemas} = $schema_cache if %schemas;
	return;
}

sub dependencies_available {
	create_doc_and_cache();
	return 1;
}

sub create_doc_and_cache {
	# Stop Win32::OLE from being noisy
	my $warn_level = Win32::OLE->Option('Warn');
	Win32::OLE->Option(Warn => 0);

	foreach my $version ('5.0', '4.0') {
		my $doc   = Win32::OLE->new('MSXML2.DOMDocument.' . $version) or next;
		my $cache = Win32::OLE->new('MSXML2.XMLSchemaCache.' . $version) or next;
		$MSXML_VERSION = $version;
		Win32::OLE->Option(Warn => $warn_level); # restore warn level
		return ($doc,$cache);
	}
	die "Unable to instantiate MSXML DOMDocument and SchemaCache. (Do you have a compatible version of MSXML installed?)";
}

# Note: Our use of TRACE and DUMP here is a bit weird. We explicitly pass to
# the TRACE and DUMP in the superclass (XML::Validate::Base) because we expect
# to be dynamically loaded and we assume that the calling class will have dealt
# with Base but not this module. (Note that Log::Trace now has some support for
# dynamic loading. It doesn't play well with some modules in 5.6.1, but it seems
# fine in 5.8. So someday this won't be necessary.)

sub TRACE { XML::Validate::Base::TRACE(@_) }
sub DUMP  { XML::Validate::Base::DUMP(@_)  }

dependencies_available();

__END__

=head1 NAME

XML::Validate::MSXML - Interface to MSXML validator

=head1 SYNOPSIS

  my $validator = new XML::Validate::MSXML(%options);
  if ($doc = $validator->validate($xml)) {
    ... Do stuff with $doc ...
  } else {
    print "Document is invalid\n";
  }

=head1 DESCRIPTION

XML::Validate::MSXML is an interface to Microsoft's MSXML parser (often
available in Windows environments) which can be used with the XML::Validate
module.

=head1 METHODS

=over

=item new(%options)

Returns a new XML::Validate::MSXML instance using the specified options. (See
OPTIONS below.)

=item validate($xml)

Returns true if $xml could be successfully parsed, undef otherwise.

=item last_dom()

Returns the MSXML DOM object of the document last validated.

=item last_error()

Returns the error from the last validate call. This is a hash ref with the
following fields:

=over

=item *

message

=item *

line

=item *

column

=back

Note that the error gets cleared at the beginning of each C<validate> call.

=item version()

Returns the version of the MSXML component that is installed

=back

=head1 OPTIONS

XML::Validate::MSXML takes the following options:

=over

=item strict_validation

If this boolean value is true, the document will be validated during parsing.
Otherwise it will only be checked for well-formedness. Defaults to true.

=back

=head1 ERROR REPORTING

When a call to validate fails to parse the document, the error may be retrieved
using last_error.

On errors not related to the XML parsing, these methods will throw exceptions.
Wrap calls with eval to catch them.

=head1 PACKAGE GLOBALS

$XML::Validate::MSXML::MSXML_VERSION contains the version number of MSXML.

=head1 DEPENDENCIES

Win32::OLE, MSXML 4.0 or 5.0

=head1 VERSION

$Revision: 1.17 $ on $Date: 2005/10/07 07:59:53 $ by $Author: johna $

=head1 AUTHOR

Nathan Carr, Colin Robertson

E<lt>cpan _at_ bbc _dot_ co _dot_ ukE<gt>

=head1 COPYRIGHT

(c) BBC 2005. This program is free software; you can redistribute it and/or modify it under the GNU GPL.
See the file COPYING in this distribution, or http://www.gnu.org/licenses/gpl.txt

=cut


lib/XML/Validate.pm
# Factory class providing a common interface to different XML validators

package XML::Validate;

# Using XML::Validate::Base here is cheating somewhat. We do this because we're
# dynamically loading the validation module, but we still want to be able to
# use Log::Trace's deep import feature. Any better ideas welcomed.

use strict;
use XML::Validate::Base;
use vars qw($VERSION);

$VERSION = sprintf"%d.%03d", q$Revision: 1.24 $ =~ /: (\d+)\.(\d+)/;

# Xerces is preferred over MSXML as we've found in practice that MSXML4's schema validation occasionally
# lets through invalid documents which Xerces catches.
# At the time of writing, LibXML didn't have schema validation support.
my $DEFAULT_PRIORITISED_LIST = [ qw( Xerces MSXML LibXML ) ];

sub new {
	my $class = shift;
	my %args = @_;
	
	DUMP("Arguments", %args);
	
	my $self = {
		backend => undef,
	};
	
	bless($self,$class);
	
	if (!$args{Type}) {
		die "Required argument Type missing\n";
	}
	
	if ($args{Type} !~ /^\w+$/) {
		die "Validator type name '$args{Type}' should only contain word characters.\n";
	}
	
	my $backend_class;
	if ($args{Type} eq 'BestAvailable') {
		my $list = $args{PrioritisedList} || $DEFAULT_PRIORITISED_LIST;
		my $type = $self->_best_available($list) or die sprintf("None of the listed backends (%s) are available\n",join(", ",@$list));
		$backend_class = "XML::Validate::" . $type;
	} else {
		$backend_class = "XML::Validate::" . $args{Type};
	}
	eval "use $backend_class";
	die "Validator $backend_class not loadable: $@" if $@;

	my $options = $args{Options};
	$self->{backend} = new $backend_class(%{$options});
	
	return $self;
}

sub validate {
	my $self = shift;
	$self->{backend}->validate(@_);
}

sub last_error {
	my $self = shift;
	$self->{backend}->last_error(@_);
}

sub type {
	my $self = shift;
	my $class = ref($self->{backend});
	$class =~ m/XML::Validate::(.*)/;
	return $1;
}

sub version {
	my $self = shift;
	return unless $self->{backend}->can('version');
	return $self->{backend}->version;
}

sub _best_available {
	my $self = shift;
	my ($list) = @_;
	foreach my $backend (@{$list}) {
		TRACE("Attempting to load $backend");
		eval "use XML::Validate::$backend";
		next if $@;
		TRACE("Loading succeeded. Returning $backend");
		return $backend;
	}
	return;
}

sub TRACE {}
sub DUMP {}

1;

__END__

=head1 NAME

XML::Validate - an XML validator factory

=head1 SYNOPSIS

  my $validator = new XML::Validate(Type => 'LibXML');
  
  if ($validator->validate($xml)) {
    print "Document is valid\n";
  } else {
    print "Document is invalid\n";
    my $message = $validator->last_error()->{message};
    my $line = $validator->last_error()->{line};
    my $column = $validator->last_error()->{column};
    print "Error: $message at line $line, column $column\n";
  }

=head1 DESCRIPTION

XML::Validate is a generic interface to different XML validation backends.
For a list of backend included with this distribution see the README.

If you want to write your own backends, the easiest way is probably to subclass
XML::Validate::Base. Look at the existing backends for examples.

=head1 METHODS

=over

=item new(Type => $type, Options => \%options)

Returns a new XML::Validate parser object of type $type. For available types see README or use 'BestAvailable' (see
L<BEST AVAILABLE>).

The optional argument Options can be used to supply a set of key-value pairs to
the backend parser. See the documentation for individual backends for details
of these options.

=item validate($xml_string)

Attempts a validating parse of the XML document $xml_string and returns a true
value on success, or undef otherwise. If the parse fails, the error can be
inspected using C<last_error>.

Note that documents which don't specify a DTD or schema will be treated as
valid.

For DOM-based parsers, the DOM may be accessed by instantiating the backend module directly and calling the C<last_dom> method - consult the documentation of the specific backend modules.
Note that this isn't formally part of the XML::Validate interface as non-DOM-based validators may added at some point.

=item last_error()

Returns the error from the last validate call. This is a hash ref with the
following fields:

=over

=item *

message

=item *

line

=item *

column

=back

Note that the error gets cleared at the beginning of each C<validate> call.

=item type()

Returns the type of backend being used.

=item version()

Returns the version of the backend

=back

=head1 ERROR REPORTING

When a call to validate fails to parse the document, the error may be retrieved using last_error.

On errors not related to the XML parsing methods will throw exceptions.
Wrap calls with eval to catch them.

=head1 BEST AVAILABLE

The BestAvailable backend type will check which backends are available and give
you the "best" of those. For the default order of preference see the README with this distribution, but this can be changed with the option PrioritisedList.

If Xerces and LibXML are available the following code will give you a LibXML backend:

  my $validator = new XML::Validate(
      Type => 'BestAvailable',
      Options => { PrioritisedList => [ qw( MSXML LibXML Xerces ) ] },
  );

=head1 KNOWN ISSUES

There is a bug in versions 1.57 and 1.58 of XML::LibXML that causes an issue
related to DTD loading. When a base parameter is used in conjunction with the
load_ext_dtd method the base parameter is ignored and the current directory
is used as the base parameter. In other words, when validating XML with LibXML
any base parameter option will be ignored, which may result in unexpected DTD
loading errors. This was reported as bug on November 30th 2005 and the bug
report can be viewed here http://rt.cpan.org/Public/Bug/Display.html?id=16213

=head1 VERSION

$Revision: 1.24 $ on $Date: 2006/04/07 12:21:08 $ by $Author: mattheww $

=head1 AUTHOR

Nathan Carr, Colin Robertson

E<lt>cpan _at_ bbc _dot_ co _dot_ ukE<gt>

=head1 COPYRIGHT

(c) BBC 2005. This program is free software; you can redistribute it and/or modify it under the GNU GPL.
See the file COPYING in this distribution, or http://www.gnu.org/licenses/gpl.txt

=cut

